code/game/objects/effects/portals.dm

```diff
@@ -28,42 +45,113 @@
 /obj/effect/portal/make_frozen_visual()
 	return
 
-/obj/effect/portal/New(loc, turf/target, creator=null, lifespan=300)
-	..()
+/obj/effect/portal/Crossed(atom/movable/AM, oldloc)
+	if(get_turf(oldloc) == get_turf(linked))
+		return ..()
+	if(!teleport(AM))
+		return ..()
+
+/obj/effect/portal/attack_tk(mob/user)
+	return
+
+/obj/effect/portal/attack_hand(mob/user)
+	if(Adjacent(user))
+		teleport(user)
+
+/obj/effect/portal/Initialize(mapload, _creator, _lifespan = 300, obj/effect/portal/_linked, automatic_link = TRUE, hard_target_override, atmos_link_override)
+	. = ..()
 	GLOB.portals += src
-	src.target = target
-	src.creator = creator
+	if(!istype(_linked) && automatic_link)
+		. = INITIALIZE_HINT_QDEL
+		CRASH("Somebody fucked up.")
+	if(_lifespan > 0)
+		QDEL_IN(src, _lifespan)
+	if(!isnull(atmos_link_override))
+		atmos_link = atmos_link_override
+	link_portal(_linked)
+	hardlinked = automatic_link
+	creator = _creator
 
-	var/area/A = get_area(target)
-	if(A && A.noteleport) // No point in persisting if the target is unreachable.
-		qdel(src)
-		return
-	if(lifespan > 0)
-		QDEL_IN(src, lifespan)
+/obj/effect/portal/proc/link_portal(obj/effect/portal/newlink)
+	linked = newlink
+	if(atmos_link)
+		link_atmos()
 
-/obj/effect/portal/Destroy()
-	GLOB.portals -= src
-	if(istype(creator, /obj/item/weapon/hand_tele))
-		var/obj/item/weapon/hand_tele/O = creator
-		O.active_portals--
-	else if(istype(creator, /obj/item/weapon/gun/energy/wormhole_projector))
-		var/obj/item/weapon/gun/energy/wormhole_projector/P = creator
-		P.portal_destroyed(src)
+/obj/effect/portal/proc/link_atmos()
+	if(atmos_source || atmos_destination)

```

Comment:
link_atmos() isn't even called much, only on creation (unless someone adds mobile portals) so this isn't exactly a costly check.